File: 907. Sum of Subarray Minimums.js
// Monotonic approach
// Time complexiity - O(N)
// Space complexity- O(2*N)
var sumSubarrayMins = function(arr) {
    const length = arr.length;
    const left = new Array(length).fill(-1);
    const right = new Array(length).fill(length);
    const stack = [];

    for (let i = 0; i < length; ++i) {
        while (stack.length > 0 && arr[stack[stack.length - 1]] >= arr[i]) {
            stack.pop();
        }
        if (stack.length) 
        left[i] = stack[stack.length - 1];
        
        stack.push(i);
    }

    stack.length = 0;
    for (let i = length - 1; i >= 0; --i) {
        while (stack.length > 0 && arr[stack[stack.length - 1]] > arr[i]) {
            stack.pop();
        }
        
        if (stack.length) 
        right[i] = stack[stack.length - 1];
        
        stack.push(i);
    }

    let sum = 0;
    const mod=1e9+7;
    for (let i = 0; i < length; ++i) {
        sum += (i - left[i]) * (right[i] - i) * arr[i] % mod;
        sum %= mod;
    }

    return sum;
};

File: 907. Sum of Subarray Minimums.cpp
// Monotonic stack
// Time complexity - O(N)
// Space complexity- O(2*N)
#define ll long long
class Solution {
    public:
    int sumSubarrayMins(vector<int>& arr) {
        int n=arr.size();
        vector<int> left_small(n,0);
        vector<int> right_small(n,0);
        
        //left small
        stack<int> st1;
        for(int i=0;i<n;i++)
        {
            while(!st1.empty() and arr[st1.top()]>=arr[i])
            st1.pop();
            
            if(st1.empty())
            left_small[i]=0;
            else
            left_small[i]=st1.top()+1;
            
            st1.push(i);
        }
        
        //right small
        stack<int> st2;
        for(int i=n-1;i>=0;i--)
        {
            while(!st2.empty() and arr[st2.top()]>arr[i])
            st2.pop();
            
            if(st2.empty())
            right_small[i]=n-1;
            else
            right_small[i]=st2.top()-1;
            
            st2.push(i);
        }
        
        
        ll ans=0;
        int mod=1e9+7;
        
        for(int i=0;i<n;i++)
        {
            int left=i-left_small[i]+1;
            int right=right_small[i]-i+1;
            ll temp_ans=(static_cast<ll> (left) * right * arr[i] )%mod;
            ans=(ans+temp_ans)%mod;
            
        }
        
        return ans;
    }
};

/*
  (left+1) * (right+1) * arr[i]
  
  consider current element as minimum and
  find the number of subarray that can be 
  form using current element , 
*/